SQL.Rmd

```diff
@@ -0,0 +1,1190 @@
+| `INTEGER` `BOOLEAN` | A whole integer value, e.g. age. Sometimes a boolean value is also represented as an integer value of `0` or `1`. |
+| `FLOAT` `DOUBLE` `REAL` | Floating point to represent fractional values. |
+| `CHARACTER(num_chars)` `VARCHAR(num_chars)` `TEXT` | Strings and text with the difference related to efficiency of the database. The types with `num_chars` specify the maximum number of characters. |
+| `DATE` `DATETIME` | Storing date and time stamps. |
+
+#### Table Constraints
+
+Staying at a high level, here are some of the constraints that can be placed on columns:
+
+| Constraint        | Description                             | 
+| :---------------: | --------------------------------------- | 
+| `PRIMARY KEY` | The values in this column are unique and can identify a single row in the table. |
+| `AUTOINCREMENT` | For integer values, the value of this will be automatically filled with a automatically incrementing value with each row insertion (not supported in all databases). |
+| `UNIQUE` | Similar to a primary key in that the values have to be unique but are not used as a key for that table. |
+| `NOT NULL` | The value cannot be `NULL`. |
+| `CHECK (expression)` | A more complex expression to test values, e.g. positive values, greater than values, starting with certain prefix, etc. |
+| `FOREIGN KEY` | Consistency check that each value in this column corresponds to another value in another table. |
+
+To test this out, let's create a new table called `smra_test`. This will return `NULL` as the query doesn't ask for anything to be returned. Until you add some data, there won't be anything to display either, we'll get to that next. Until then, let's look at our columns:
+
+* `id` - this is just going to be an automatically increasing integer but will be used to uniquely identify the row.
+* `first_name` - while it shouldn't be null, it also shouldn't be longer than 50 characters.
+* `specialty` - this alphanumeric code will not be longer than 4 characters.
+* `los` - (length of stay) is a positive integer.
+
+```{sql create-table, connection="conn", exercise=TRUE}
+
+```
+
+```{sql create-table-solution}
+CREATE TABLE IF NOT EXISTS smra_test (
+  id INTEGER PRIMARY KEY AUTOINCREMENT,
+  first_name CHARACTER(50) NOT NULL,
+  specialty CHARACTER(4),
+  los INTEGER
+);
+```
+
+</br>
+
+### INSERT INTO
+
+When inserting data into a database, we use the `INSERT` statement followed by the name of the table to write to, the columns of data we are filling, and one or more rows of data to insert. If you are relying on default values and not listing every column you need to state the column names explicitly. It's also possible to utilise expressions for the values to insert. 
+
+```sql
+INSERT INTO my_table
+-- optional explicit column listing
+(some_column, another_column, ...)
+VALUES (some_value, another_value...)...;
+```
+
+In our created table, `smra_test`, let's add a couple of rows:
```

Comment:
This doesn't work for me, if I run the code in the solution I get an error saying smra_test doesn't exist and when I add the code to create the table it returns NULL. I tried using a query to pull back the data from smra_test and still got NULL.